### PR TITLE
Scan recency filter for older recency pages

### PR DESCRIPTION
**Pull request by FiV0**
This adds the recency filter to scan for older pages than the query range of scan. 

The formula is essentially `while recency of page < min(validTo.lower, systemTo.lower)` we are ignoring those pages. 

Todo:
- [x] Walk through scan recency test.

### CODE REVIEW

File: core/src/main/kotlin/xtdb/trie/HashTrie.kt [hunk 1]
```diff
@@ -97,11 +101,13 @@ fun toMergePlan(segments: List<ISegment>, pathPred: Predicate<ByteArray>?): List
...
                 for (mpNode in mergePlanTask.mpNodes) {
-                    // TODO filter by recencies #3166
                     val recencies = mpNode.node.recencies
                     if (recencies != null) {
-                        for (i in recencies.indices) {
+                        var i = 0
+                        while (i < recencies.size && recencies[i] <= minRecency) i++
+                        while (i < recencies.size) {
                             newMpNodes += MergePlanNode(mpNode.segment, mpNode.node.recencyNode(i))
+                            i++
```
Comment (FiV0) [inline core/src/main/kotlin/xtdb/trie/HashTrie.kt]: This 5 lines are actually the logic, so probably good to have a second :eyes: over this.

File: src/test/clojure/xtdb/operator/scan_test.clj [hunk 2]
```diff
@@ -676,3 +679,56 @@
...
+                       (count (tu/query-ra '[:scan {:table docs :for-valid-time [:between #inst "2025" nil]} [xt$id]]
+                                           {:node node}))))
+              ;; 2024 and inf
+              (t/is (= 2 @merge-tsk-cnt))
+
+
+              ;; everything
+              (t/is (= 5
+                       (count (tu/query-ra '[:scan {:table docs :for-valid-time :all-time} [xt$id]] {:node node}))
+                       @merge-tsk-cnt)))))))))
```
Comment (FiV0) [inline src/test/clojure/xtdb/operator/scan_test.clj]: This test is a bit of a mouthful (or however you would say that for a test). So maybe we can find something simpler.

File: core/src/main/kotlin/xtdb/util/TemporalBounds.kt [hunk 3]
```diff
@@ -20,11 +21,19 @@ class TemporalBounds {
         fun gte(operand: Long) = apply { lower = max(operand, lower) }
 
         fun inRange(operand: Long) = operand in lower..upper
+
+        override fun toString(): String {
```
Comment (jarohen) [inline core/src/main/kotlin/xtdb/util/TemporalBounds.kt]: these might be better achieved by making TemporalBounds and TemporalColumn data classes, if that works?

Comment (FiV0) [reply core/src/main/kotlin/xtdb/util/TemporalBounds.kt]: Can make them data classes. The override for `TemporalColumn` still needs to happen if I want a human friendlier output or is there a way to annotate data fields with special printers?

File: src/test/clojure/xtdb/operator/scan_test.clj [hunk 4]
```diff
@@ -676,3 +679,57 @@
...
+          (tu/finish-chunk! node)
+          ;; compaction happens in 2025
+          (c/compact-all! node)
+
+          (let [->old-merge-tasks scan/->merge-tasks
+                merge-tsk-cnt (atom 0)
+                query-opts {:node node
+                            :basis {:at-tx tx-key
+                                    :current-time (:system-time tx-key)}}]
+            (with-redefs [scan/->merge-tasks (fn [& args]
```
Comment (jarohen) [inline src/test/clojure/xtdb/operator/scan_test.clj:705]: as discussed, investigate whether we can call `->merge-tasks` as a unit test here instead

File: src/test/clojure/xtdb/trie_test.clj [hunk 5]
```diff
@@ -29,39 +29,78 @@
...
+                                           (-> (trie/->Segment (->arrow-hash-trie log-root))
+                                               (assoc :seg :log))
+                                           (-> (trie/->Segment (->arrow-hash-trie log2-root))
+                                               (assoc :seg :log2))]
+                                          nil
+                                          (TemporalBounds.))
+                  (merge-plan-nodes->path+pages))))))
+
+(deftest test-merge-plan-recency-filtering
+  (with-open [al (RootAllocator.)
```
Comment (jarohen) [inline src/test/clojure/xtdb/trie_test.clj:73]: todos:

- [x] comment up exactly what the test case is stressing - what's the case, what are you hoping the algo will do?
- [x] are there other cases to consider here which might catch the algo out?

File: src/test/clojure/xtdb/operator/scan_test.clj (hunk 2 above)
Comment (jarohen) [reply src/test/clojure/xtdb/operator/scan_test.clj]: as discussed, let's remove the with-redefs here, just test the final answer, and rely on the unit tests below to confirm that it got there in the correct manner.

value of these tests is in showing that, even for cases which might cause the algo trouble, we still return the correct answers - i.e. we haven't missed a page

**Review by jarohen — APPROVED** [review]
nice one @FiV0 :clap:

**Pull request merged**